trame_server/client.py

```diff
@@ -0,0 +1,302 @@
+                    if msg_id == WsLinkSession.AUTH_ID:
+                        self.client_id = msg_result.get("clientID")
+                        future.set_result(self.client_id)
+                    else:
+                        future.set_result(msg_result)
+
+                # Clean pending future
+                if future:
+                    self.in_flight_rpc.pop(msg_id)
+
+            elif msg.type == aiohttp.WSMsgType.BINARY:
+                if len(self.in_flight_attachment):
+                    key = self.in_flight_attachment.pop(0)
+                    self.attachments[key] = msg.data
+                else:
+                    raise ValueError("Attachment missing registration key")
+            elif msg.type == aiohttp.WSMsgType.CLOSE:
+                print("CLOSE")
+            elif msg.type == aiohttp.WSMsgType.CLOSING:
+                print("CLOSING")
+            elif msg.type == aiohttp.WSMsgType.CLOSED:
+                print("CLOSED")
+            elif msg.type == aiohttp.WSMsgType.ERROR:
+                print("ERROR")
+                break
+
+    async def auth(self, **kwargs):
+        key = WsLinkSession.AUTH_ID
+        resp = self.loop.create_future()
+        msg = dict(
+            wslink="1.0",
+            id=key,
+            method="wslink.hello",
+            args=[kwargs],
+            kwargs={},
+        )
+        self.in_flight_rpc[key] = resp
+        await self.ws.send_str(json.dumps(msg))
+        return resp
+
+    async def call(self, method, args=None, kwargs=None):
+        self.msg_count += 1
+        key = f"rpc:{self.client_id}:{self.msg_count}"
+        resp = self.loop.create_future()
+        self.in_flight_rpc[key] = resp
+        if args is None:
+            args = []
+        if kwargs is None:
+            kwargs = {}
+
+        await self.ws.send_str(
+            json.dumps(
+                dict(
+                    wslink="1.0",
+                    id=key,
+                    method=method,
+                    args=args,
+                    kwargs=kwargs,
+                )
+            )
+        )
+
+        return resp
+
+    def register_subscription(self, topic, callback):
+        if topic not in self.subscriptions:
+            self.subscriptions[topic] = [callback]
+        else:
+            self.subscriptions[topic].append(callback)
+
+    def unregister_subscription(self, topic, callback):
+        callbacks = self.subscriptions.get(topic, [])
+        if callback in callbacks:
+            callbacks.remove(callback)
+
+        if len(callbacks) == 0 and topic in self.subscriptions:
+            self.subscriptions.pop(topic)
+
+    def clear_subscriptions(self):
+        topics = list(self.subscriptions.keys())
+        for topic in topics:
+            self.subscriptions.pop(topic)
+
+
+class Client:
+    """
+    Client implementation for driving a remote trame server with its shared state and trigger method calls in plain python.
+    """
```

Comment:
It could be nice to have a base class shared with the `Server` where identical methods are located to reduce repetition, but not necessary.

Reply:
Not sure, since everything underneath is different. To some extent if we were typing thing, we would need an interface.